### ISSUE TITLE
Releases after 3.1.0 fail on private key extraction with passphrase 

### ISSUE THREAD

**Issue by axu562** [mattosaurus/PgpCore#122]
Beginning with version 3.1.0, attempting to decrypt a file with DecryptFileAsync (including a non-empty passphrase) results in a Checksum mismatch error in Org.BouncyCastle.
The file is able to be decrypted through other methods, so it is not an issue with the file, secret key, or passphrase itself.
In addition, we upgraded bouncy castle to its latest version while using Pgp version 2.2.0, and it decrypted the file properly. This seems to only be an issue with PGPCore later versions. The issue seems to have begun with version 3.1.0, and is still failing on the latest version.

**Comment by mattosaurus** [comment]
Weird. The tests are all running as expected for me, but entirely possible they don't cover everything.

https://github.com/mattosaurus/PgpCore/blob/58d946fedc9c63512a39eb01fa6bcb92cb9eaf52/PgpCore.Tests/UnitTests/LegacyUnitTestsAsync.cs#L252-L275

Are you encrypting the files using PgpCore as well or just using it to decrypt files encrypted elsewhere?

If you're able to provide me with a set of keys and files that I can use to replicate this I'll have a look into it.

**Comment by axu562** [comment]
After a little more digging, I think this only occurs with .Z files. When either encrypting a .Z file or decrypting a .Z.gpg file, we get a Checksum mismatch at 0 of 2 error. I've attached a stacktrace of the exception (this is for encrypting a .Z file with a public key, private key, and passphrase).
I'm also working on generating a .Z and keys that will cause this failure, but I suspect any .Z file and non-empty passphrase or keys should do it.
![trace](https://user-images.githubusercontent.com/40678478/114757586-0996d800-9d2a-11eb-9f53-851d48f699f3.png)

**Comment by axu562** [comment]
So unfortunately I cannot seem to recreate this with another .Z file. The files that are having issues are generated by clients, and so we don't know exactly what they did to cause this to fail. A plain .Z file seems to decrypt and encrypt with no issues.
It does seem like maybe this has something to do with the encoding of the passphrase? We see similar errors if the passphrase is wrong, but in this case since we are able to decrypt with other methods the passphrase is definitely correct. Perhaps later versions are somehow not grabbing the correct algorithm / encoding for the passphrase.
Thanks for looking into this anyway, and sorry that I can't provide better examples!

**Comment by malhotrar** [comment]
So I work with @axu562 and I spent some time checking out the code and diving into this. I believe the difference is that between v2.2.0 and 3.1.0+ EncryptionKeys are now always configured despite not all methods in PGP actually needing all properties of EncryptionKeys being set.  

If I run the "ReadPrivateKey(passphrase)" method in 2.2.0 with our private key and passphrase I get the same exact error as we are seeing above in later versions, so our issue stems that from that in 2.2.0 simply running DecryptFileAsync() with any version of BouncyCastle.Portable results in a success because the ReadPrivateKey method where the error would be thrown is never invoked. 

Am I correct in my assumption that not all properties in EncryptionKeys are actually required for uses within the PGP class? If so, would it make sense for the properties to be Lazy such that they are only instantiated when required?

**Comment by mattosaurus** [comment]
It certainly makes sense that this would be happening due to the changes made to the EncryptionKeys object.

You're correct that depending on the method not all EncryptionKeys properties are required. However the ReadPrivateKey method just uses the private key and passphrase properties which should have been supplied in order to decrypt the file. This runs as expected for me when I populate an instance of EncryptionKeys with public key, private key and passphrase.

```
EncryptionKeys encryptionKeys = new EncryptionKeys(testFactory.PublicKeyFileInfo, testFactory.PrivateKeyFileInfo, testFactory.Password);
encryptionKeys.ReadPrivateKey(testFactory.Password); // Success
```

Is the above essentially what you're running in your test which is failing? If so does this fail with newly generated keys from PgpCore or gpg.exe?

Are you able to submit a PR with a fix or if not a replicable example (without your live keys)?

Making the EncryptionKeys properties Lazy is probably a good idea anyway but it would be good to confirm this actually fixes your issue before pushing out a new release.

**Comment by malhotrar** [comment]
Exactly - that block is throwing a checksum mismatch 0 of 2 inside BouncyCastle ```private PgpPrivateKey ReadPrivateKey(string passPhrase)```

![image](https://user-images.githubusercontent.com/11787307/114947232-bd779080-9e1a-11eb-91b1-0cd35de578a3.png)

succeeds when I do the following:

![image](https://user-images.githubusercontent.com/11787307/114947258-c9fbe900-9e1a-11eb-8475-231a38d865fe.png)

So lazy loading would absolutely help, as those properties aren't required for decrypt file.

I was able to find a backup asc with the public and private key on a folder than someone from operations created when this keyset was created in 2011. and I believe I may be able to fix the problem by generating a new private key using the following (at least locally):

```gpg --no-default-keyring --keyring=E:\temp\pubring.gpg --export-secret-keys KEY > secring.gpg```

**Comment by mattosaurus** [comment]
So, I've made the EncryptionKeys properties Lazy but I've got no way to test if this resolves your issue or not.

https://github.com/mattosaurus/PgpCore/tree/lazy-encryption-keys

Are you able to test using this branch?

**Comment by malhotrar** [comment]
@mattosaurus  - this works perfectly! Thanks for such a quick turnaround!

**Comment by mattosaurus** [comment]
Nice, I've published v5.2.0 with this fix to NuGet so it should hopefully appear soon.
